CDAP-17981: Add opnHttpURLConnection to master env runnable context interface

wyzhang opened the pull request:
Why:
Allow master env runnable to access CDAP internal services.
Default implementation opens a connection to appfab internal REST endpoints.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/master/environment/DefaultMasterEnvironmentRunnableContext.java:
```diff
@@ -16,21 +16,45 @@
+                                                 LocationFactory locationFactory) {
+    this.discoveryServiceClient = discoveryServiceClient;
     this.locationFactory = locationFactory;
+    this.remoteClientInternal = new RemoteClient(discoveryServiceClient, Constants.Service.APP_FABRIC_HTTP,
```
Is it always be app-fabric?

wyzhang replied:
I suppose appfab will handle all user-facing and extension-facing requests in the future, thus default to appfab.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/master/environment/DefaultMasterEnvironmentRunnableContext.java:
```diff
@@ -16,21 +16,45 @@
     this.locationFactory = locationFactory;
+    this.remoteClientInternal = new RemoteClient(discoveryServiceClient, Constants.Service.APP_FABRIC_HTTP,
+                                                 new DefaultHttpRequestConfig(false),
+                                                 String.format("%s", Constants.Gateway.INTERNAL_API_VERSION_3));
```
So it can only call internal api?

wyzhang replied:
After a second thought, no need to restrict to internal. Removed.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/master/environment/DefaultMasterEnvironmentRunnableContextTest.java:
```diff
@@ -0,0 +1,122 @@
+ */
+public class DefaultMasterEnvironmentRunnableContextTest {
+  @ClassRule
+  public static final TemporaryFolder TEMP_FOLDER = new TemporaryFolder();
```
Unused? You should use it to setup the cConf to use temp folder for this test (by using `LocalLocationModule` as well). Temp directory can be set via the `Constants.CFG_LOCAL_DATA_DIR` config.

wyzhang replied:
Removed. No need at the moment. 

Thanks for the pointer. Will keep it mind. Needed for a followup PR.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/master/environment/DefaultMasterEnvironmentRunnableContext.java:
```diff
@@ -16,21 +16,45 @@
+   */
+  @Override
+  public HttpURLConnection openHttpURLConnection(String resource) throws IOException {
+    return remoteClientInternal.openConnection(resource);
```
Do we need to use `HttpsEnabler` if the connection is an instance of `HttpsURLConnection`?

wyzhang replied:
RemoteClient already takes care of that. 

io.cdap.cdap.common.internal.remote.RemoteClient#resolve figures out if the target service runs with http or https and then constructs either a corresponding URL. Then httpsEnabler skips the cert validation by setting always-trusting on the connection as below. 

  public HttpURLConnection openConnection(String resource) throws IOException {
    URL url = resolve(resource);
    HttpURLConnection urlConn = (HttpURLConnection) url.openConnection();
    if (urlConn instanceof HttpsURLConnection && !httpRequestConfig.isVerifySSLCert()) {
      new HttpsEnabler().setTrustAll(true).enable((HttpsURLConnection) urlConn);
    }

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/master/environment/DefaultMasterEnvironmentRunnableContext.java:
```diff
@@ -16,21 +16,44 @@
+  private final RemoteClient remoteClient;
 
-  public DefaultMasterEnvironmentRunnableContext(LocationFactory locationFactory) {
+  @Inject
```
Seems like it is only used in test? If that's the case, better not to use Guice.

wyzhang replied:
Used in MasterEnvironmentMain.   No need for Guice. Removed.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/master/environment/DefaultMasterEnvironmentRunnableContextTest.java:
```diff
@@ -0,0 +1,118 @@
+          bind(DiscoveryServiceClient.class).toInstance((DiscoveryServiceClient) discoveryService);
+        }
+      });
+    context = injector.getInstance(DefaultMasterEnvironmentRunnableContext.class);
```
Not really need to use Guice. Just new the `LocalLocationFactory` with a base path pointing to a test directory created via `TemporaryFolder` is better.

wyzhang replied:
Done

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/master/environment/DefaultMasterEnvironmentRunnableContextTest.java:
```diff
@@ -0,0 +1,118 @@
+    String resource = String.format("echo/%s", message);
+    HttpURLConnection conn = context.openHttpURLConnection(resource);
+    conn.setRequestMethod(HttpMethod.GET.name());
+    InputStream is = conn.getInputStream();
```
This is leaking resource. Better just use `HttpRequests` class to help writing the test?

wyzhang replied:
Wrapped in a try with resource. 

For using HttpRequests, none of its API takes a connection. Creating, opening and using connection are all private and within the HttpRequests.

wyzhang committed:
CDAP-17981: Add opnHttpURLConnection to master env runnable context interface.

Why:
Allow master env runnable to access CDAP internal services.
Default implementation opens a connection to appfab internal REST endpoints.
